<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title></title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:tankard@mac2172.local" />
</head>

<body>



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#REQUIREMENTS">REQUIREMENTS</a></li>
  <li><a href="#USAGE">USAGE</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#SEE-ALSO">SEE ALSO</a></li>
  <li><a href="#COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Bio::SNP::TOPBOT Determine the TOP/BOT strand</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code>    use Bio::SNP::TOPBOT; 

    # Finding strand of SNP rs11804171 with a reference sequence.
    # Requires installation of samtools in users PATH
    $reference_file = &#39;hg19.fa&#39;; # reference sequence file
    $chrom = &#39;chr1&#39;;
    $position = 723819;
    $allele_a = &#39;A&#39;;
    $allele_b = &#39;T&#39;;
    topbot_genome $reference_file, $chrom, $position, $allele_a, $allele_b;

    # Finding strand of SNP rs11804171, SNP is assumed to be in the middle.
    $sequence = &#39;AAAGTACAA&#39;;
    $allele_a = &#39;A&#39;;
    $allele_b = &#39;T&#39;;
    topbot_sequence $sequence, $allele_a, $allele_b;

    # With the location of the SNP in the sequence given. 
    $sequence = &#39;AGAAAGTACAA&#39;;
    $position = 7; 
    topbot_sequence $sequence, $allele_a, $allele_b, $position;</code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>For a SNP, determines whether the reference forward strand is the TOP or BOT strand as defined by Illumina. This scheme of strand designation is described in the <a href="http://res.illumina.com/documents/products/technotes/technote_topbot.pdf">&quot;TOP/BOT&quot; Strand and &quot;A/B&quot; Allele</a> white paper released by Illumina.</p>

<h1 id="REQUIREMENTS">REQUIREMENTS</h1>

<p>This module&#39;s <code>topbot_genome</code> depends on the SAMtools software being installed on the system.</p>

<h1 id="USAGE">USAGE</h1>

<p>Bio::SNP::TOPBOT can determine the TOP/BOT strand of a SNP on the reference forward strand with either a reference sequence file or the reference sequence directly.</p>

<p>Both <code>topbot_genome</code> and <code>topbot_sequence</code> give similar outputs. When the strand can be successfully determined the output will be &#39;TOP&#39; or &#39;BOT&#39;. When there is an error in determining the sequence then a string starting with &#39;ERROR&#39; will be returned.</p>

<p>For <code>topbot_sequence</code> only, when there is insufficient information to determine the strand (such as running out of sequence) then the empty string &#39;&#39; will be returned.</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Rick Tankard &lt;trickytank@gmail.com&gt;.</p>

<h1 id="SEE-ALSO">SEE ALSO</h1>

<h1 id="COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</h1>

<p>Copyright (c) Rick Tankard</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself, either Perl version 5.10.0 or, at your option, any later version of Perl 5 you may have available.</p>


</body>

</html>


